### PR TITLE
[DI] Allow dumping the container in one file instead of many files

### PR DESCRIPTION
**Pull request by lyrixx**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | -
| Tests pass?   | yes    
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Replaced by #32581

### CODE REVIEW

File: src/Symfony/Component/HttpKernel/Kernel.php [hunk 1]
```diff
@@ -835,7 +836,7 @@ protected function getContainerBuilder()
      * @param string           $class     The name of the class to generate
      * @param string           $baseClass The name of the container's base class
      */
-    protected function dumpContainer(ConfigCache $cache, ContainerBuilder $container, $class, $baseClass)
+    protected function dumpContainer(ConfigCache $cache, ContainerBuilder $container, $class, $baseClass, bool $asFiles = true)
```
Comment (stof) [inline src/Symfony/Component/HttpKernel/Kernel.php]: is this ever passed as `false` ?

Comment (lyrixx) [reply src/Symfony/Component/HttpKernel/Kernel.php]: if you change the $asFile that are hardcoded... yes :) I just fixed the other side of an hardcoded path :joy:

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 2]
```diff
@@ -269,6 +269,8 @@ public function dump(array $options = [])
             foreach ($this->generateProxyClasses() as $c) {
                 $code .= $c;
             }
+
+            $code .= sprintf("\nreturn new %s();\n", $options['class']);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: this should be removed (moved to the dumpContainer method)

File: src/Symfony/Component/HttpKernel/Kernel.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/Kernel.php]: Instead of an argument, this should be a parameter to me.
That would fix the BC break and make this easier to configure I believe.

**Review by nicolas-grekas — COMMENTED** [review]
That's for 4.4, it's a new feature :)

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php (hunk 2 above)
Comment (lyrixx) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: why. like that, it's consistent with the "asFiles=true" mode!

File: src/Symfony/Component/HttpKernel/Kernel.php (hunk 1 above)
Comment (lyrixx) [reply src/Symfony/Component/HttpKernel/Kernel.php]: a DIC parameter you mean?

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/Kernel.php]: yes

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: because there are ppl out there that use the component standalone an don't want a container to be created and destroyed immediately :)

Comment (lyrixx) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: Nice! So this is a bug fix !

File: src/Symfony/Component/HttpKernel/Kernel.php (hunk 1 above)
Comment (lyrixx) [reply src/Symfony/Component/HttpKernel/Kernel.php]: Fixed

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php (hunk 2 above)
Comment (lyrixx) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: Fixed

**Review by lyrixx — COMMENTED** [review]
Ok, I have reworked this PR. there was a bug (did not worked in env test), now it's :+1: 
Bonus: Now it's only affect the kernel, not the DIC \o/

File: src/Symfony/Component/HttpKernel/Kernel.php [hunk 3]
```diff
@@ -572,18 +572,25 @@ protected function getContainerBaseClass()
...
+
+                return;
+            }
             // Silence E_WARNING to ignore "include" failures - don't use "@" to prevent silencing fatal errors
             $errorLevel = error_reporting(\E_ALL ^ \E_WARNING);
             $fresh = $oldContainer = false;
             try {
                 if (file_exists($cache->getPath()) && \is_object($this->container = include $cache->getPath())) {
                     $this->container->set('kernel', $this);
-                    $oldContainer = $this->container;
```
Comment (lyrixx) [inline src/Symfony/Component/HttpKernel/Kernel.php]: This var was useless, because at the new line we have `$fresh=true` and later `if ($fresh} { return; }`

**Comment by lyrixx** [comment]
The diff is easier to read thank to [`?w=1`](https://github.com/symfony/symfony/pull/32119/files?w=1)

File: src/Symfony/Component/HttpKernel/Kernel.php [hunk 4]
```diff
@@ -844,27 +863,33 @@ protected function dumpContainer(ConfigCache $cache, ContainerBuilder $container
             $dumper->setProxyDumper(new ProxyDumper());
         }
 
+        $asFiles = $container->hasParameter('container.as_files') ? $container->getParameter('container.as_files') : true;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Kernel.php]: container.dump_as_files?
or
container.dumper.as_files?

Comment (lyrixx) [reply src/Symfony/Component/HttpKernel/Kernel.php]: Actually, I wanted to stick the current format

https://github.com/symfony/symfony/blob/64eaf7e3a63e6586f18582ea4ab8e5f334560525/src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php#L298-L300

Should I change? If so, I prefer `container.dumper.as_files`

**Comment by lyrixx** [comment]
Are you OK with this one?

File: src/Symfony/Component/HttpKernel/Kernel.php (hunk 4 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/Kernel.php]: Double checking, `container.dumper.*` is for configuring the dumper,
while this one configures how the kernel uses the dumper...
So I'm for `container.dump_as_files`, which is the most explicit.

File: src/Symfony/Component/HttpKernel/Kernel.php [hunk 5]
```diff
@@ -572,18 +572,25 @@ protected function getContainerBaseClass()
      */
     protected function initializeContainer()
     {
+        $asFiles = true;
         $class = $this->getContainerClass();
         $cacheDir = $this->warmupDir ?: $this->getCacheDir();
         $cache = new ConfigCache($cacheDir.'/'.$class.'.php', $this->debug);
         $oldContainer = null;
         if ($fresh = $cache->isFresh()) {
+            if (!$asFiles) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Kernel.php]: this cannot happen

File: src/Symfony/Component/HttpKernel/Kernel.php [hunk 6]
```diff
@@ -572,18 +572,25 @@ protected function getContainerBaseClass()
...
+                require_once $cache->getPath();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Kernel.php]: moving this outside of the try/catch+require_cone where it was before means this will fail badly when the file is missing (will happen on race conditions)

File: src/Symfony/Component/HttpKernel/Kernel.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/Kernel.php]: unless the line above throws an exception, which is why this code is how it is

File: src/Symfony/Component/HttpKernel/Kernel.php [hunk 7]
```diff
@@ -634,6 +641,13 @@ protected function initializeContainer()
         try {
             $container = null;
             $container = $this->buildContainer();
+            if ($asFiles) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Kernel.php]: always true here - but why is this needed? it's not right now

File: src/Symfony/Component/HttpKernel/Kernel.php [hunk 8]
```diff
@@ -634,6 +641,13 @@ protected function initializeContainer()
...
+                $container->getParameterBag()->add([
+                    'container.build_id' => 'static',
+                    'container.build_hash' => 'static',
+                    'container.as_files' => true,
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Kernel.php]: why is this needed? this should already be dumped as part of the regular container *when the param is defined*

File: src/Symfony/Component/HttpKernel/Kernel.php [hunk 9]
```diff
@@ -657,7 +671,12 @@ protected function initializeContainer()
         $oldContainer = \is_object($oldContainer) ? new \ReflectionClass($oldContainer) : false;
 
         $this->dumpContainer($cache, $container, $class, $this->getContainerBaseClass());
-        $this->container = require $cache->getPath();
+        if ($asFiles) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Kernel.php]: always true also, or I'm missing something

**Review by nicolas-grekas — COMMENTED** [review]
still for 4.4 to me, sorry about that :)

File: src/Symfony/Component/HttpKernel/Kernel.php (hunk 5 above)
Comment (lyrixx) [reply src/Symfony/Component/HttpKernel/Kernel.php]: :) If I change the value line 575, it will happen.
But OK, I will add something to be able to change it in a clean manier. So 4.4, you win!
But I don't know how to do that with the DIC. I think the only solution is to add a new paratemer (the DIC is not build at this time)

File: src/Symfony/Component/HttpKernel/Kernel.php (hunk 6 above)
Comment (lyrixx) [reply src/Symfony/Component/HttpKernel/Kernel.php]: This code was present in symfony 2.X + there is not try/catch block here. So I don't get what you mean

File: src/Symfony/Component/HttpKernel/Kernel.php (hunk 3 above)
Comment (lyrixx) [reply src/Symfony/Component/HttpKernel/Kernel.php]: :+1:

File: src/Symfony/Component/HttpKernel/Kernel.php (hunk 7 above)
Comment (lyrixx) [reply src/Symfony/Component/HttpKernel/Kernel.php]: Same answer, I will add a new param

File: src/Symfony/Component/HttpKernel/Kernel.php (hunk 8 above)
Comment (lyrixx) [reply src/Symfony/Component/HttpKernel/Kernel.php]: This is needed to avoid adding an extra param to the method `dumpContainer()`

File: src/Symfony/Component/HttpKernel/Kernel.php (hunk 4 above)
Comment (lyrixx) [reply src/Symfony/Component/HttpKernel/Kernel.php]: okay, I will change this

File: src/Symfony/Component/HttpKernel/Kernel.php (hunk 3 above)
Comment (lyrixx) [reply src/Symfony/Component/HttpKernel/Kernel.php]: I don't understand: 
* If the line above throw an exception, then we can not execute this line.
* If the line above does not throw exception, so we execute this line. But this line can not fail. So we execute the next line: `$fresh=true` and later `if ($fresh} { return; }`. So to me this is dead code.

File: src/Symfony/Component/HttpKernel/Kernel.php (hunk 6 above)
Comment (lyrixx) [reply src/Symfony/Component/HttpKernel/Kernel.php]: I moved that in the catch block

File: src/Symfony/Component/HttpKernel/Kernel.php (hunk 8 above)
Comment (lyrixx) [reply src/Symfony/Component/HttpKernel/Kernel.php]: Fixed

**Comment by lyrixx** [comment]
OK, I have added a test, and addressed your comments.
I also added an extra commit where I simplified the code. (about fresh / oldContainer) May be I miss something, but I think you are wrong. Anyway, It's an extra commit, so I would remove it easily! 
So it could be better to review the first commit first

**Comment by lyrixx** [comment]
I rebased this PR. It's now ready and green :green_heart:

**Comment by nicolas-grekas** [comment]
I pushed a completely different implementation on your branch: instead of patching the Kernel (I kept your cleanup), this now patches the PhpDumper. This allows keeping the anonymous class proxy, which is desired to have safe cache clears.

**Comment by nicolas-grekas** [comment]
Replaced by #32581
Thanks @lyrixx for raising the issue and giving it a try.

**Comment by lyrixx** [comment]
@nicolas-grekas  What about the cleaning ?
